### PR TITLE
Add unknown metadata field validation tests

### PR DESCRIPTION
**Pull request by kevindelgado**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Adds test coverage for how client-side validation catches unknown metadata fields because we are currently lacking coverage. This lack of coverage was exposed when server-side validation failed to detect unknown metadata fields for CRDs. The e2e tests added should work for both the current client-side usage of kubectl validation and the eventual server-side

#### Which issue(s) this PR fixes:

See action item 3 from https://github.com/kubernetes/kubernetes/issues/109215#issuecomment-1087037419

### CODE REVIEW

**Comment by kevindelgado** [comment]
Ready for review, no rush though
PTAL @liggitt, @apelisse

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -1332,7 +1332,7 @@ func (v *unstructuredSchemaCoercer) apply(u *unstructured.Unstructured) (unknown
 			if v.returnUnknownFieldPaths {
 				pruneOpts.ReturnPruned = true
 			}
-			unknownFieldPaths = structuralpruning.PruneWithOptions(u.Object, v.structuralSchemas[gv.Version], false, pruneOpts)
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:1335]: It seems that this PR changes the product code while description says it adds tests. Please clarify - update description or add a note why this change is needed

**Comment by liggitt** [comment]
this is building on https://github.com/kubernetes/kubernetes/pull/109268, only the last commit is new

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/109268 is merged, this is good to rebase

File: test/e2e/kubectl/kubectl.go [hunk 2]
```diff
@@ -1079,6 +1098,109 @@ metadata:
...
+      	          "apiVersion": "foo/v1",
+      	 	  "kind": "Sub",
+      		  "metadata": {
+			%s
+      		        "name": "subobject"
+      		  },
+      		  "spec": null
+          }
+  }
+}`
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1148]: nit: indentation/whitespace on this is mixing tabs and spaces and is pretty hard to read... indent with 2-spaces consistently?

File: test/e2e/kubectl/kubectl.go [hunk 3]
```diff
@@ -1079,6 +1098,109 @@ metadata:
...
+{%s,
+  "spec": {
+          "template": {
+      	          "apiVersion": "foo/v1",
+      	 	  "kind": "Sub",
+      		  "metadata": {
+			%s
+      		        "name": "subobject"
+      		  },
+      		  "spec": null
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1145]: why pass an explicit `null` for a non-nullable optional field?

File: test/e2e/kubectl/kubectl.go [hunk 4]
```diff
@@ -102,6 +102,7 @@ const (
 	httpdDeployment2Filename  = "httpd-deployment2.yaml.in"
 	httpdDeployment3Filename  = "httpd-deployment3.yaml.in"
 	metaPattern               = `"kind":"%s","apiVersion":"%s/%s","metadata":{"name":"%s"}`
+	unknownMetaPattern        = `"kind":"%s","apiVersion":"%s/%s","metadata":{"unknownMeta": "foo", "name":"%s"}`
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:105]: I know this is a pre-existing pattern, but it makes for almost unreadable tests ... seems like a generator function like `func unknownFieldMetadataJSON(gvk schema.GroupVersionKind, name string) string { ... }` would be more readable

File: test/e2e/kubectl/kubectl.go [hunk 5]
```diff
@@ -163,6 +164,24 @@ properties:
...
+var schemaFooEmbedded = []byte(`description: Foo CRD with an embedded resource
+type: object
+properties:
+  spec:
+    type: object
+    properties:
+      template:
+        type: object
+        x-kubernetes-embedded-resource: true
+        properties:
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:176]: I can't remember if explicitly defining a `metadata` property with a `name` subproperty made a difference or not in the parent issue... might consider doing that here, then exercising an implicit field (`namespace`), an explicit field (`name`), and an unknown field (`unknownField`) when writing the custom resource

Comment (kevindelgado) [reply test/e2e/kubectl/kubectl.go:176]: Discussed offline, I now have the CRD explicitly defining the `name` subproperty within the `metadata `property for both root and embedded objects. I’m also passing both `name` and `namespace` values for root and embedded `metadata` fields on all requests to confirm that these are handled properly.

For now, unknown fields are passed for root and embedded metadata objects separately until I can find a better way to validate multiple error messages that is consistent both now (with the client performing the unknown field validation) and after unknown field validation is performed default on the server.

**Review by kevindelgado — COMMENTED** [review]
Feedback addressed, ptal @liggitt

**Comment by kevindelgado** [comment]
CI is all green now, ptal @liggitt

File: test/e2e/kubectl/kubectl.go [hunk 6]
```diff
@@ -1079,6 +1106,111 @@ metadata:
...
+      }
+    }
+  }
+}`
+			meta := unknownFieldMetadataJSON(gvk, "test-cr")
+			unknownRootMetaCR := fmt.Sprintf(embeddedCRPattern, meta, "")
+			_, err = framework.RunKubectlInput(ns, unknownRootMetaCR, "create", "--validate=true", "-f", "-")
+			if !(strings.Contains(err.Error(), `unknown field "unknownMeta"`) || strings.Contains(err.Error(), `unknown field "metadata.unknownMeta"`)) {
+				framework.Failf("error missing root unknown metadata field, got: %v", err)
+			}
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1163]: also assert the error does not complain about the namespace field

File: test/e2e/kubectl/kubectl.go [hunk 7]
```diff
@@ -1079,6 +1106,111 @@ metadata:
...
+
+			ginkgo.By("attempting to create a CR with unknown metadata fields in the embedded object")
+			metaEmbedded := fmt.Sprintf(metaPattern, testCRD.Crd.Spec.Names.Kind, testCRD.Crd.Spec.Group, testCRD.Crd.Spec.Versions[0].Name, "test-cr-embedded")
+			unknownEmbeddedMetaCR := fmt.Sprintf(embeddedCRPattern, metaEmbedded, `"unknownMetaEmbedded": "bar",`)
+			_, err = framework.RunKubectlInput(ns, unknownEmbeddedMetaCR, "create", "--validate=true", "-f", "-")
+			if !(strings.Contains(err.Error(), `unknown field "unknownMetaEmbedded"`) || strings.Contains(err.Error(), `unknown field "spec.template.metadata.unknownMetaEmbedded"`)) {
+				framework.Failf("error missing embedded unknown metadata field, got: %v", err)
+			}
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1171]: separately assert the error does not complain about the namespace field

File: test/e2e/kubectl/kubectl.go [hunk 8]
```diff
@@ -1079,6 +1106,111 @@ metadata:
...
+				}
+			}
+		}
+	}
+		`
+			_, err := framework.RunKubectlInput(ns, invalidMetaDeployment, "create", "-f", "-")
+			if !(strings.Contains(err.Error(), `unknown field "unknownMeta"`) || strings.Contains(err.Error(), `unknown field "metadata.unknownMeta"`)) {
+				framework.Failf("error missing unknown metadata field, got: %v", err)
+			}
+
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1212]: separately assert the error does not complain about the namespace field

File: test/e2e/kubectl/kubectl.go [hunk 9]
```diff
@@ -1079,6 +1106,111 @@ metadata:
...
+					"containers": [{
+						"name":  "nginx",
+						"image": "nginx:latest"
+					}]
+				}
+			}
+		}
+	}
+		`
+			_, err := framework.RunKubectlInput(ns, invalidMetaDeployment, "create", "-f", "-")
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1208]: before calling err.Error(), assert err is non-nil (to avoid an unclear panic error)

File: test/e2e/kubectl/kubectl.go [hunk 10]
```diff
@@ -1079,6 +1106,111 @@ metadata:
...
+			unknownRootMetaCR := fmt.Sprintf(embeddedCRPattern, meta, "")
+			_, err = framework.RunKubectlInput(ns, unknownRootMetaCR, "create", "--validate=true", "-f", "-")
+			if !(strings.Contains(err.Error(), `unknown field "unknownMeta"`) || strings.Contains(err.Error(), `unknown field "metadata.unknownMeta"`)) {
+				framework.Failf("error missing root unknown metadata field, got: %v", err)
+			}
+
+			ginkgo.By("attempting to create a CR with unknown metadata fields in the embedded object")
+			metaEmbedded := fmt.Sprintf(metaPattern, testCRD.Crd.Spec.Names.Kind, testCRD.Crd.Spec.Group, testCRD.Crd.Spec.Versions[0].Name, "test-cr-embedded")
+			unknownEmbeddedMetaCR := fmt.Sprintf(embeddedCRPattern, metaEmbedded, `"unknownMetaEmbedded": "bar",`)
+			_, err = framework.RunKubectlInput(ns, unknownEmbeddedMetaCR, "create", "--validate=true", "-f", "-")
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1168]: before calling err.Error(), assert err is non-nil (to avoid an unclear panic error)

File: test/e2e/kubectl/kubectl.go [hunk 11]
```diff
@@ -1079,6 +1106,111 @@ metadata:
...
+        %s
+        "name": "subobject",
+        "namespace": "my-ns"
+      }
+    }
+  }
+}`
+			meta := unknownFieldMetadataJSON(gvk, "test-cr")
+			unknownRootMetaCR := fmt.Sprintf(embeddedCRPattern, meta, "")
+			_, err = framework.RunKubectlInput(ns, unknownRootMetaCR, "create", "--validate=true", "-f", "-")
```
Comment (liggitt) [inline test/e2e/kubectl/kubectl.go:1160]: before calling err.Error(), assert err is non-nil (to avoid an unclear panic error)

**Review by kevindelgado — COMMENTED** [review]
feedback addressed, PTAL @liggitt
